File: example/src/main/java/com/github/vivchar/example/pages/simple/ViewStateFragment.java
package com.github.vivchar.example.pages.simple;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.vivchar.example.BaseScreenFragment;
import com.github.vivchar.example.R;
import com.github.vivchar.example.widgets.BetweenSpacesItemDecoration;
import com.github.vivchar.rendererrecyclerviewadapter.CompositeViewHolder;
import com.github.vivchar.rendererrecyclerviewadapter.CompositeViewState;
import com.github.vivchar.rendererrecyclerviewadapter.DefaultCompositeViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer;
import com.github.vivchar.rendererrecyclerviewadapter.ViewState;
import com.github.vivchar.rendererrecyclerviewadapter.binder.CompositeViewBinder;
import com.github.vivchar.rendererrecyclerviewadapter.binder.CompositeViewStateProvider;
import com.github.vivchar.rendererrecyclerviewadapter.binder.ViewBinder;

import java.util.Collections;
import java.util.List;

/**
 * Created by Vivchar Vitaly on 28.12.17.
 */

public class ViewStateFragment extends BaseScreenFragment {

	private final YourDataProvider mYourDataProvider = new YourDataProvider();
	private RendererRecyclerViewAdapter mRecyclerViewAdapter;
	@Nullable
	private Bundle mSavedInstanceState;

	@Nullable
	@Override
	public View onCreateView(final LayoutInflater inflater,
	                         @Nullable final ViewGroup container,
	                         @Nullable final Bundle savedInstanceState) {

		final View view = inflater.inflate(R.layout.fragment_list, container, false);

		mRecyclerViewAdapter = new RendererRecyclerViewAdapter();

		mRecyclerViewAdapter.registerRenderer(
				new CompositeViewBinder<>(
						R.layout.item_simple_composite,
						R.id.recycler_view,
						StateViewModel.class,
						Collections.singletonList(new BetweenSpacesItemDecoration(10, 10)),
						new CompositeViewStateProvider<StateViewModel, CompositeViewHolder>() {
							@Override
							public ViewState createViewState(@NonNull final CompositeViewHolder holder) {
								return new CompositeViewState(holder);
							}

							@Override
							public int createViewStateID(@NonNull final StateViewModel model) {
								return model.getID();
							}
						}
				).registerRenderer(getAnyViewRenderer())
		);
//		adapter.registerRenderer(...);
//		adapter.registerRenderer(...);
		mRecyclerViewAdapter.setItems(mYourDataProvider.getStateItems());

		final RecyclerView recyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
		recyclerView.setAdapter(mRecyclerViewAdapter);
		recyclerView.addItemDecoration(new BetweenSpacesItemDecoration(10, 10));

		return view;
	}

	@Override
	public void onViewStateRestored(@Nullable final Bundle savedInstanceState) {
		super.onViewStateRestored(savedInstanceState);
		mSavedInstanceState = savedInstanceState;
	}

	@Override
	public void onResume() {
		super.onResume();
		mRecyclerViewAdapter.onRestoreInstanceState(mSavedInstanceState);
		mSavedInstanceState = null;
	}

	@Override
	public void onSaveInstanceState(@NonNull final Bundle outState) {
		super.onSaveInstanceState(outState);
		mRecyclerViewAdapter.onSaveInstanceState(outState);
	}

	private ViewRenderer getAnyViewRenderer() {
		return new ViewBinder<>(R.layout.item_simple_square, DiffUtilFragment.DiffViewModel.class,
				(model, finder, payloads) -> finder.setText(R.id.text, model.getText())
		);
	}

	public static class StateViewModel extends DefaultCompositeViewModel {

		private final int mID;

		public StateViewModel(final int ID, @NonNull final List<? extends ViewModel> items) {
			super(items);
			mID = ID;
		}

		private int getID() {
			return mID;
		}
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/ViewStateFragment.kt
package com.github.vivchar.example.pages.simple

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.github.vivchar.example.R
import com.github.vivchar.example.base.BaseFragment
import com.github.vivchar.example.databinding.FragmentListBinding
import com.github.vivchar.example.pages.simple.DiffUtilFragment.DiffViewModel
import com.github.vivchar.example.widgets.BetweenSpacesItemDecoration
import com.github.vivchar.example.widgets.MyAdapter
import com.github.vivchar.rendererrecyclerviewadapter.*
import kotlinx.coroutines.launch

class ViewStateFragment : BaseFragment<FragmentListBinding>() {

	private val viewModel: ViewStateViewModel by viewModels()
	private val adapter = MyAdapter()
	private var instanceState: Bundle? = null

	override fun createBinding(inflater: LayoutInflater, container: ViewGroup?) =
		FragmentListBinding.inflate(inflater, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		val compositeViewRenderer = CompositeViewRenderer(
			R.layout.item_simple_composite,
			R.id.recycler_view,
			StateViewModel::class.java,
			listOf(BetweenSpacesItemDecoration(10, 10)),
			object : CompositeViewStateProvider<StateViewModel?, CompositeViewHolder<ViewFinder>> {
				override fun createViewState() = CompositeViewState<CompositeViewHolder<*>>()
				override fun createViewStateID(model: StateViewModel) = model.id
			}
		)
		compositeViewRenderer.registerRenderer(
			ViewRenderer<DiffViewModel, ViewFinder>(
				R.layout.item_simple_square,
				DiffViewModel::class.java
			) { (_, text), finder, _ -> finder.setText(R.id.text, text) }
		)

		adapter.registerRenderer(compositeViewRenderer)
		binding.recyclerView.adapter = adapter
		binding.recyclerView.addItemDecoration(BetweenSpacesItemDecoration(10, 10))

		viewLifecycleOwner.lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				viewModel.state.collect { state ->
					adapter.setItems(state.items)
				}
			}
		}
	}

	override fun onViewStateRestored(savedInstanceState: Bundle?) {
		super.onViewStateRestored(savedInstanceState)
		instanceState = savedInstanceState
	}

	override fun onResume() {
		super.onResume()
		adapter.onRestoreInstanceState(instanceState)
		instanceState = null
	}

	override fun onSaveInstanceState(outState: Bundle) {
		super.onSaveInstanceState(outState)
		adapter.onSaveInstanceState(outState)
	}

	data class StateViewModel(val id: Int, val list: List<com.github.vivchar.rendererrecyclerviewadapter.ViewModel>) : DefaultCompositeViewModel(list)
}
